libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -17,14 +24,142 @@ type KadDHT* = ref object of LPProtocol
+    try:
+      let reply = await replyFut
+      receivedReplies.add(reply)
+    except CatchableError:
+      failedPeers.add(peerId)
+      error "could not send find_node to peer", peerId, err = getCurrentExceptionMsg()
+
+  return (receivedReplies, failedPeers)
+
+proc findNode*(
+    kad: KadDHT, targetId: Key
+): Future[seq[PeerId]] {.async: (raises: [CancelledError]).} =
+  #debug "findNode", target = target
+  # TODO: should it return a single peer instead? read spec
+
+  var initialPeers = kad.rtable.findClosestPeers(targetId, DefaultReplic)
+  var state = LookupState.init(targetId, initialPeers)
+  var addrTable: Table[PeerId, seq[MultiAddress]] =
+    initTable[PeerId, seq[MultiAddress]]()
+
+  while not state.done:
+    let toQuery = state.selectAlphaPeers()
+
+    var pendingFutures = initTable[PeerId, Future[Message]]()
+
+    for peer in toQuery:
+      if pendingFutures.hasKey(peer):
+        continue
+
+      state.markPending(peer)
+
+      pendingFutures[peer] = kad
+        .sendFindNode(peer, addrTable.getOrDefault(peer, @[]), targetId)
+        .wait(5.seconds)
+
+      state.activeQueries.inc
+
+    let (successfulReplies, timedOutPeers) = await waitRepliesOrTimeouts(pendingFutures)
+
+    for msg in successfulReplies:
+      for peer in msg.closerPeers:
+        addrTable[PeerId.init(peer.id).get()] = peer.addrs
+      state.updateShortlist(
+        msg,
+        proc(p: PeerInfo) =
+          discard kad.rtable.insert(p.peerId)
+          kad.switch.peerStore[AddressBook][p.peerId] = p.addrs
+          # TODO: add TTL to peerstore, otherwise we can spam it with junk
+        ,
+      )
+
+    for timedOut in timedOutPeers:
+      state.markFailed(timedOut)
+
+    state.done = state.checkConvergence()
+
+  return state.selectClosestK()
+
+proc bootstrap*(
+    kad: KadDHT, bootstrapNodes: seq[PeerInfo]
+) {.async: (raises: [CancelledError]).} =
+  for b in bootstrapNodes:
+    try:
+      await kad.switch.connect(b.peerId, b.addrs)
+      debug "connected to bootstrap peer", peerId = b.peerId
+
+      try:
+        let msg =
+          await kad.sendFindNode(b.peerId, b.addrs, kad.rtable.selfId).wait(5.seconds)
+        for peer in msg.closerPeers:
+          let p = PeerId.init(peer.id).tryGet()
+          discard kad.rtable.insert(p)
+          kad.switch.peerStore[AddressBook][p] = peer.addrs
+
+        # bootstrap node replied succesfully. Adding to routing table
+        discard kad.rtable.insert(b.peerId)
+      except CatchableError as e:
+        error "bootstrap failed for peer", peerId = b.peerId, exc = e.msg
+    except CatchableError as e:
+      error "failed to connect to bootstrap peer", peerId = b.peerId, error = e.msg
+
+  try:
+    # Adding some random node to prepopulate the table
+    discard await kad.findNode(PeerId.random(kad.rng).tryGet().toKey())
+    info "bootstrap lookup complete"
+  except CatchableError as e:
+    error "bootstrap lookup failed", error = e.msg
```

Comment:
Continuing our conversation about your find about code not following specs:
> On every run, we generate a random peer ID for every non-empty routing table's k-bucket 

We can probably fix this by replacing this code by a call to `refreshBuckets` which is supposed to do what spec suggests.
```suggestion
await refreshBuckets()
```

And remove `maintainBuckets` proc,

Also, `refreshBuckets` will likely not need that `if kad.rtable.buckets[i].isStale():` that checks if the bucket contains stale peers or not, and can just execute findNode directly.

Reply:
sounds good